crates/mdb/src/lib.rs

```diff
@@ -151,61 +95,75 @@ impl Drop for Connection {
         // to it at construction so accessing `on_error` without synchronization is safe.
         unsafe {
             mdb_sys::mdb_connection_destroy(&mut self.ptr);
-            drop(Box::from_raw(self.on_error));
         }
     }
 }
 
 unsafe impl Send for Connection {}
 unsafe impl Sync for Connection {}
 
+struct Deferred(Option<Box<dyn FnOnce()>>);
+impl Drop for Deferred {
+    fn drop(&mut self) {
+        assert!(self.0.is_some());
+        self.0.take().unwrap()()
```

Comment:
Don't we need to call the function here?

`new` takes a pointer and creates a function that, when called, creates a box then drops that box. Now, if we never call the function, then the box is never created and never dropped.